### ISSUE TITLE
Installing/running on Azure cloud

### ISSUE THREAD

**Issue by eeeehaw** [facebook/fbctf#219]
I've been trying to get fbctf running on an Azure VM that is running Ubuntu 14.04.  I can ssh into the machine remotely and have setup http and https endpoints on Azure, but when attempting to access fbctf from a browser nginx sends an error "502 Bad Gateway".  I'm thinking it may be related to fbctf being hard coded to work on IPaddr 10.10.10.5, instead of the IPaddress assigned by DHCP to the machine??  Or is there another possible cause?

FBCTF is installed on the Azure Ubuntu VM in "production" configuration/mode, so Vagrant is not installed (I have confirmed Vagrant is not installed), therefore changing the private IPaddress in the Vagrantfile has NO effect.  How can I change the IPaddr that fbctf uses for http/https etc? 

It would be useful to publish an architecture description of the fbctf system so users know what each program is doing (eg, is nginx being used as the webserver, or proxy, or load balancer, or some combination of those?)

Here's some info captured from the Ubuntu console terminal for that Azure Ubuntu server VM (note that the Azure Endpoint for ssh/http/https/et al points to the VM's DHCP assigned IP address, which may change between VM start/stop sessions):
--------------
login as: azureuser
azureuser@cyberwars.cloudapp.net's password:
Welcome to Ubuntu 14.04.5 LTS (GNU/Linux 4.4.0-38-generic x86_64)
   __               _                 _     _____ _______ ______
  / _|             | |               | |   / ____|__   __|  ____|
 | |_ __ _  ___ ___| |__   ___   ___ | | _| |       | |  | |__
 |  _/ _' |/ __/ _ \  _ \ / _ \ / _ \| |/ / |       | |  |  __|
 | || (_| | (_|  __/ |_) | (_) | (_) |   <| |____   | |  | |
 |_| \__,_|\___\___|_.__/ \___/ \___/|_|\_\\_____|  |_|  |_|

  System information as of Wed Sep 21 02:04:16 UTC 2016

  System load: 0.16              Memory usage: 6%   Processes:       97
  Usage of /:  6.1% of 28.80GB   Swap usage:   0%   Users logged in: 0

  Graph this data and manage this system at:
    https://landscape.canonical.com/

  Get cloud support with Ubuntu Advantage Cloud Guest:
    http://www.ubuntu.com/business/services/cloud

New release '16.04.1 LTS' available.
Run 'do-release-upgrade' to upgrade to it.

Your Hardware Enablement Stack (HWE) is supported until April 2019.

Last login: Tue Sep 20 17:48:54 2016 from c-73-19-22-162.hsd1.wa.comcast.net

azureuser@cyberwars:~$ ifconfig
eth0      Link encap:Ethernet  HWaddr 00:0d:3a:f7:27:0e
          inet addr:100.104.72.32  Bcast:100.104.75.255  Mask:255.255.252.0
          inet6 addr: fe80::20d:3aff:fef7:270e/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1040 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1268 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:357664 (357.6 KB)  TX bytes:186026 (186.0 KB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

azureuser@cyberwars://var/log$ curl -4ivk https://100.104.72.32
* Rebuilt URL to: https://100.104.72.32/
* Hostname was NOT found in DNS cache
*   Trying 100.104.72.32...
* Connected to 100.104.72.32 (100.104.72.32) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS handshake, Server key exchange (12):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using ECDHE-RSA-AES128-GCM-SHA256
* Server certificate:
*        subject: O=Facebook CTF
*        start date: 2016-09-20 18:01:56 GMT
*        expire date: 2017-09-20 18:01:56 GMT
*        issuer: O=Facebook CTF
*        SSL certificate verify result: self signed certificate (18), continuing anyway.
> GET / HTTP/1.1
> User-Agent: curl/7.35.0
> Host: 100.104.72.32
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
HTTP/1.1 502 Bad Gateway
* Server nginx is not blacklisted
< Server: nginx
Server: nginx
< Date: Wed, 21 Sep 2016 18:09:58 GMT
Date: Wed, 21 Sep 2016 18:09:58 GMT
< Content-Type: text/html
Content-Type: text/html
< Content-Length: 166
Content-Length: 166
< Connection: keep-alive
Connection: keep-alive

<
<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
* Connection #0 to host 100.104.72.32 left intact
azureuser@cyberwars://var/log$

------------------------

**Comment by iangcarroll** [comment]
What's the output of `hhvm --version`? And is there anything in `/var/log/nginx/error.log` and `/var/log/hhvm/error.log`?

**Comment by eeeehaw** [comment]
(1) hhvm --version  ==>  "HipHop VM 3.14.5(rel)
(2) /var/log/hhvm/error.log is empty (nothing in it)
(3) /var/log/nginx/error.log contains the following:

2016/09/21 18:06:23 [crit] 1447#0: *18 connect() to 
unix:/var/run/hhvm/sock failed (2: No such file or directory) while      
                            connecting to upstream, client: 
100.104.72.32, server: , request: "GET / HTTP/1.1", upstream: 
"fastcgi://unix:/var/run/h                                hvm/sock:", 
host: "100.104.72.32"
2016/09/21 18:09:58 [crit] 1447#0: *20 connect() to 
unix:/var/run/hhvm/sock failed (2: No such file or directory) while      
                            connecting to upstream, client: 
100.104.72.32, server: , request: "GET / HTTP/1.1", upstream: 
"fastcgi://unix:/var/run/h                                hvm/sock:", 
host: "100.104.72.32"
2016/09/21 20:44:39 [warn] 1491#0: "ssl_stapling" ignored, issuer 
certificate not found
2016/09/21 20:56:44 [crit] 1496#0: *5 connect() to 
unix:/var/run/hhvm/sock failed (2: No such file or directory) while c    
                             onnecting to upstream, client: 73.19.22.162, 
server: , request: "GET / HTTP/1.1", upstream: 
"fastcgi://unix:/var/run/hhv                                m/sock:", 
host: "cyberwars.cloudapp.net"
2016/09/21 20:56:44 [error] 1496#0: *5 open() 
"/var/www/fbctf/src/favicon.ico" failed (2: No such file or directory), 
cl                                ient: 73.19.22.162, server: , request: 
"GET /favicon.ico HTTP/1.1", host: "cyberwars.cloudapp.net", referrer: 
"https://c                                yberwars.cloudapp.net/"
2016/09/21 20:56:44 [crit] 1496#0: *5 connect() to 
unix:/var/run/hhvm/sock failed (2: No such file or directory) while c    
                             onnecting to upstream, client: 73.19.22.162, 
server: , request: "GET /favicon.ico HTTP/1.1", upstream: 
"fastcgi://unix:/                                var/run/hhvm/sock:", 
host: "cyberwars.cloudapp.net", referrer: 
"https://cyberwars.cloudapp.net/"

(4)  When I curl https://10.10.10.5 from console of Ubuntu VM, it 
doesn't respond ie is taking so long listening for a response I have to 
ctrl-z out of that command.  Not surprising, of course, since the VM is 
not in a 10.0.0.0 subnet on Azure at the moment, so the VM eth0 can't 
respond.  I found out that there is supposedly a way to create a virtual 
network with private subnet that includes host IPaddr 10.10.10.5 within 
Azure, but to accomplish that I have to delete everything I've done so 
far and start over (grrr...).  Surely there is a place within the 
production installation version of fbctf to change the private address 
from 10.10.10.5 to either one in another specified network, or else have 
it dynamically adjust to DHCP issued address to the localhost VM (the 
better solution)??

thx, Stephen

**Comment by iangcarroll** [comment]
I believe HHVM is just having trouble starting. Does the directory /var/run/hhvm exist? Is the HHVM service running (`service hhvm status`)? If it's not starting, try deleting everything in /var/run/hhvm.

**Comment by eeeehaw** [comment]
Ian, you're correct!  hhvm is not running (see below) and I can't 
restart it to running result.  What should I try next?

Yes, /var/run/hhvm exists.

Moving the files to a subdirectory "old" I created and trying to restart 
the hhvm service results in these clues (ie, still would not result in 
hhvm running):
--------------
azureuser@cyberwars:/var/run/hhvm$ ls
hhvm.hhbc  pid
azureuser@cyberwars:/var/run/hhvm$ service hhvm status
  * hhvm is not running
azureuser@cyberwars:/var/run/hhvm$ sudo mkdir old
azureuser@cyberwars:/var/run/hhvm$ sudo mv hhvm.hhbc ./old/hhvm.hhbc.old
azureuser@cyberwars:/var/run/hhvm$ sudo mv pid ./old/pid.old
azureuser@cyberwars:/var/run/hhvm$ ls
old
azureuser@cyberwars:/var/run/hhvm$ sudo service hhvm restart
  * Restarting HHVM FastCGI Daemon hhvm                            [ OK ]
azureuser@cyberwars:/var/run/hhvm$ service hhvm status
  * hhvm is not running
azureuser@cyberwars:/var/run/hhvm$ ls
hhvm.hhbc  old  pid
azureuser@cyberwars:/var/run/hhvm$

---------------------

**Comment by iangcarroll** [comment]
Do you see anything in `/var/log/hhvm/error.log` now? Or in any log file inside `/var/log/hhvm`? 

If not, make sure HHVM can write to that directory and files.

**Comment by eeeehaw** [comment]
No, both the error.log and access.lot are empty, or more specifically 
there are about 50 blank lines.

Here's the /var/log/hhvm directory attributes:
drwxrwxr-x 2 www-data  www-data    4096 Sep 20 17:55 hhvm

Here's the file permission attributes:
azureuser@cyberwars:/var/log/hhvm$ ls -l
total 0
-rw-rw-rw- 1 root     root     0 Sep 20 17:55 access.log
-rw-rw-r-- 1 www-data www-data 0 Aug 16 01:33 error.log

Looks like HHVM should already have access/write privileges to this IF 
it's the Owner in a system Group?  (I see that the Other section has 
only read permission; I'm not aware of how HHVM is permitted in the 
Ubuntu system)  If not, how can I fix it?

THX Ian for your ongoing attention on this, highly appreciated by me.

--Stephen

**Comment by eeeehaw** [comment]
So, I'm wondering: since the fbctf is configured for server address of 10.10.10.5, and the Azure server virtual machine is on an entirely different subnet, perhaps there is some check/loopback on the server's eth0 that one or more modules are doing in fbctf that fails, resulting in HHVM not running?  Ie, I'm still wondering where I can change that IPaddr in fbctf to match the Azure VM machine IP?

**Comment by gsingh93** [comment]
Sorry I haven't had time to help you with your actual issue, but I just wanted to chime in here and say FBCTF does absolutely *no* configuration of the IP address. 10.10.10.5 is simply the default _if_ you use the provided Vagrantfile. Otherwise, it's whatever was configured on your server.

**Comment by eeeehaw** [comment]
thx for chiming in gsingh93

So if I'm off-base chasing the 10.10.10.5 IPaddr issue, ie if there's no other config file or compiled binary that is configured for that private localhost IP, then Ian has me on a worthy track of finding out why HHVM is refusing to run its service.  I'm don't yet have a good idea how to troubleshoot that.

So, how is NGINX being used in FBCTF, ie what's it's architectural role?  Proxy & filter only?  Load balancing?  Or as the webserver in lieu of Appache?  The answer to why I'm seeing a 502 Bad Gateway error from nginx seems dependent upon the answer to that question according to what I've found so far in my google search.

**Comment by gsingh93** [comment]
NGINX is being used as the webserver. You're seeing the bad gateway message because NGINX is trying to communicate with HHVM over a socket, but HHVM isn't running for some reason (i.e. it crashes/exits when you try to run it).

There are two things I've seen cause this. The first is permissions errors on the socket (in /var/run/hhvm/ off the top of my head), and the second is that you need to regenerate the HHVM byte code that's used in repo-authoritative mode (instructions here: https://docs.hhvm.com/hhvm/advanced-usage/repo-authoritative).

**Comment by eeeehaw** [comment]
Problem is solved.
I deleted my entire Azure Resource Group (virtual machine, virtual network, storage, and cloud resources) and started over.  This time in the new Resource Group I setup a private virtual network subnet with class A 10.10.10.0/24 address pace, assigned static IP address 10.10.10.5 to my Ubuntu 14.04 virtual machine, installed FBCTF and things are now working, ie can access the website with no 502 or other errors, logged into FBCTF as admin now.  HHVM service shows as running.  As far as I can tell so far, the only thing different in this Azure Resource Group is the address range of the vnet and static IPaddr for the VM being in the 10.10.10.0/24 space.  I recommend that future Azure users of FBCTF do the same to insure success.
Ian and gsingh93:  THX so much for your response and diagnostic help, I learned a few things from you guys so time not wasted on your part afaic.  If you need further details on this from me, holler.

**Comment by eeeehaw** [comment]
Is there a writeup somewhere that I can use to complete setting up my environment  for CTFs?  I'm not sure what to do next after getting FBCTF platform running on Auzre (I also installed it on Google Cloud BTW, but I've noticed a couple usability advantages of Azure, plus I get Azure for free use in a school setting).  I guess I had imagined that the FBCTF VM would be a target, but that doesn't make sense regarding CTF contestants not allowed by rules to attack the FBCTF infrastructure, so I'm imagining I'll need to set up additional VMs as attack targets (and machines for students to configure defense against attacks).  

Ie, where can I find a primer to give me the big picture and for next steps?

Meanwhile, I'll try to close this problem thread given that my original post is solved.  Thanks Ian and Gulshan for your help on that!
